File: backend/app/services/seathashmap_service.py
from cpp_backend import SeatHashMap
import asyncio

# Singleton instance of SeatHashMap
seat_hash_map = SeatHashMap()
lock = asyncio.Lock()

async def add_seat(seat_id: int, is_available: bool, seat_type: str):
    async with lock:
        # Check if seat already exists
        try:
            seat_hash_map.getSeatInfo(seat_id)
            # If no exception, seat exists
            return {"success": False, "error": "Seat already exists"}
        except:
            # Seat does not exist, safe to add
            seat_hash_map.addSeat(seat_id, is_available, seat_type)
    return {"success": True}

async def is_seat_available(seat_id: int):
    async with lock:
        available = seat_hash_map.isSeatAvailable(seat_id)
    return {"success": True, "available": available}

async def update_seat_availability(seat_id: int, is_available: bool):
    async with lock:
        seat_hash_map.updateSeatAvailability(seat_id, is_available)
    return {"success": True, "message": f"Seat {seat_id} availability updated"}

async def get_seat_info(seat_id: int):
    async with lock:
        try:
            seat_info = seat_hash_map.getSeatInfo(seat_id)
            return {
                "success": True,
                "seatID": seat_info.seatID,
                "isAvailable": seat_info.isAvailable,
                "seatType": seat_info.seatType,
            }
        except Exception as e:
            return {"success": False, "error": str(e)}

async def get_all_seats():
    async with lock:
        seats = seat_hash_map.getAllSeats()
        # seats is a list of SeatInfo objects
        seat_list = []
        for seat in seats:
            seat_list.append({
                "seatID": seat.seatID,
                "isAvailable": seat.isAvailable,
                "seatType": seat.seatType
            })
    return {"success": True, "seats": seat_list}

File: frontend/src/services/seatHashMapService.js
import apiClient from "./apiClient";

// Add a new seat to the hash map
export const addSeat = async (seatId, isAvailable, seatType) => {
  try {
    const response = await apiClient.post("/seathashmap/add", {
      seat_id: seatId,
      is_available: isAvailable,
      seat_type: seatType,
    });
    return response.data; // response includes success or error
  } catch (error) {
    const errMsg =
      error.response?.data?.error ||
      error.response?.data?.detail ||
      "Failed to add seat.";
    throw new Error(errMsg);
  }
};

// Check if a seat is available
export const isSeatAvailable = async (seatId) => {
  const response = await apiClient.get("/seathashmap/available", {
    params: { seat_id: seatId },
  });
  return response.data;
};

// Update the availability of a seat
export const updateSeatAvailability = async (seatId, isAvailable) => {
  const response = await apiClient.put("/seathashmap/update", {
    seat_id: seatId,
    is_available: isAvailable,
  });
  return response.data;
};

// Get information about a specific seat
export const getSeatInfo = async (seatId) => {
  try {
    const response = await apiClient.get("/seathashmap/info", {
      params: { seat_id: seatId },
    });
    return response.data;
  } catch (error) {
    const errMsg =
      error.response?.data?.error ||
      error.response?.data?.detail ||
      "Failed to get seat info.";
    throw new Error(errMsg);
  }
};

// Get all seats in the hash map
export const getAllSeats = async () => {
  try {
    const response = await apiClient.get("/seathashmap/all");
    return response.data; // returning entire data {success: true, seats: [...]}
  } catch (error) {
    throw new Error(
      error.response?.data?.detail || "Error fetching seat data.",
    );
  }
};
